File: tutorials/maximum_intensity_projection/matlab/script_mip.m
% script to compute a digitally reconstructed radiography (DRR) using
% maximum intensity projection (MIP)

%% load a 3D CT volume
load('../data/test_lung_ct.mat')

%% get the DRR position
% x -> anterior->posterior
% y -> right->left (drr_x)
% z -> superior->caudal (drr_y)

% - using x-ray source coordinates, i.e. x0=[0,0,0], for spherical coordinates
% - assuming 100 cm distance between source to detector
% - assuming a drr_size cm^2 detector with a resolution of drr_voxdims under the body
d_x2d = 1000;
drr_size = [200,120];  % [y,x] in DRR
drr_voxdims = [1,1];

% centring the volume coordinates
vol_size = size(vol);
[vol_x, vol_y, vol_z] = meshgrid(d_x2d - (vol_size(2)-0.5:-1:0.5)*voxdims(2), ...
                                (-vol_size(1)/2+0.5:vol_size(1)/2-0.5)*voxdims(1), ...
                                (-vol_size(3)/2+0.5:vol_size(3)/2-0.5)*voxdims(3)); % volume on the detector
vol_ds = sqrt(vol_x.^2+vol_y.^2+vol_z.^2);

% detector:
[drr_x,drr_y] = meshgrid((-drr_size(2)/2+0.5:drr_size(2)/2-0.5) * drr_voxdims(2), ...
                        (-drr_size(1)/2+0.5:drr_size(1)/2-0.5) * drr_voxdims(1));
drr_ds = sqrt(drr_x.^2 + drr_y.^2 + d_x2d^2);
% get the range of r
r_max = max([vol_ds(:);drr_ds(:)]);
r_min = min([vol_ds(:);drr_ds(:)]);
n_samples = ceil(1.5*(r_max-r_min));
% get spehrical coordinates, [az,el,r]
[az,el] = cart2sph(drr_x,drr_y,d_x2d);
az = repmat(az,[1,1,n_samples]);
el = repmat(el,[1,1,n_samples]); 
r = repmat(reshape(linspace(r_min,r_max,n_samples),1,1,n_samples),[drr_size(1),drr_size(2),1]);
% convert back to cartesian
[sample_z, sample_y, sample_x] = sph2cart(az,el,r);
% figure, plot3(sample_x(:),sample_y(:),sample_z(:),'.'); axis equal

% interpolation get sample values
samples = interp3(vol_x,vol_y,vol_z,single(vol),sample_x,sample_y,sample_z,'*linear',0);

% compute DDR
DRR = max(samples,[],3);
%% display
figure, imshow(DRR,[])

File: tutorials/maximum_intensity_projection/python/script_mip.py
# script to compute a digitally reconstructed radiography (DRR) using
# maximum intensity projection (MIP)

from scipy.io import loadmat
import numpy as np
from matplotlib import pyplot as plt
from scipy.interpolate import interpn

## load a 3D CT volume data
# 1-j -> anterior->posterior
# 0-i -> right->left (drr_x)
# 2-k -> superior->caudal (drr_y)
mat_data = loadmat('../data/test_lung_ct.mat')
vol = mat_data['vol'].astype('float16')  # plt.imshow(vol[:100,:,100])
voxdims = mat_data['voxdims'][0]

# - using x-ray source coordinates, i.e. x0=[0,0,0], for spherical coordinates
# - assuming 100 cm distance between source to detector
# - assuming a drr_size cm^2 detector with a resolution of drr_voxdims under the body
d_x2d = 1000
drr_size = (200,120)
drr_voxdims = (0.8,0.8)


## centring the volume coordinates
vol_i = np.linspace(-vol.shape[0]/2+0.5,vol.shape[0]/2-0.5,vol.shape[0])*voxdims[0]
vol_j = d_x2d - np.linspace(vol.shape[1]-0.5,0.5,vol.shape[1])*voxdims[1]
vol_k = np.linspace(-vol.shape[2]/2+0.5,vol.shape[2]/2-0.5,vol.shape[2])*voxdims[2]


## detector:
[drr_i, drr_j] = np.meshgrid(
    np.linspace(-drr_size[0]/2+0.5,drr_size[0]/2-0.5,drr_size[0])*drr_voxdims[0],
    np.linspace(-drr_size[1]/2+0.5,drr_size[1]/2-0.5,drr_size[1])*drr_voxdims[1],
    indexing='ij')


# get the range of r
vol_ds = np.sqrt(sum([x**2 for x in np.meshgrid(vol_i,vol_j,vol_k,indexing='ij')]))
drr_ds = np.sqrt(drr_i**2+drr_j**2+d_x2d**2)
r_max = max([vol_ds.max(),drr_ds.max()])
r_min = min([vol_ds.min(),drr_ds.min()])
n_samples = int(np.ceil(1.5*(r_max-r_min)))
# get spehrical coordinates
az = np.arctan2(drr_i,drr_j)[...,np.newaxis]
el = np.arctan2(d_x2d,np.sqrt(drr_i**2 + drr_j**2))[...,np.newaxis]
r = np.reshape(np.linspace(r_min,r_max,n_samples),(1,1,n_samples))
# convert back to cartesian
sample_z = r * np.cos(el) * np.cos(az)
sample_y = r * np.cos(el) * np.sin(az)
sample_x = r * np.sin(el)


## interpolation get sample values
samples = interpn(
                (vol_i,vol_j,vol_k), 
                vol, 
                np.stack([sample_y,sample_x,sample_z],axis=3),
                method='linear', bounds_error=False, fill_value=0.0,
                )

# compute DDR
DRR = np.amax(samples,axis=2)


## display
# plt.imshow(DRR,cmap='gray')
# plt.show()

## save to a file
plt.imsave('../data/drr_python.png',DRR,cmap='gray')
